### ISSUE TITLE
uid reported without username in samplers is problematic 

### ISSUE THREAD

**Issue by baallan** [ovis-hpc/ovis#1055]
It has been brought to our attention by a user wanting to do analyses across multiple systems that UIDs may be different for the same username and that containerized use (with uids being mapped to in-container values that differ) may also be problematic.
They have requested that we include both username and uid, at least optionally.
This affects the /proc/$pid related samplers at minimum.

**Comment by baallan** [comment]
As the fix is simple enough and lack of it constitutes a usability bug for our analysis systems, I will create a patch for it 11/3-11/4 in time for 4.3.10 inclusion.

**Comment by tom95858** [comment]
Hi @baallan, I think this might be a bit of a partial fix. Could you please provide a more complete use-case. For example is this analysis being done from the store? How do we handle the converse where the uid is the same, but the username is different? For example, totucke, vs. tom for uid 1234. I certainly wouldn't want to rush a change like this into a release under test without full understanding what the goals/side-effects are.

**Comment by baallan** [comment]
@tom95858 My proposed solution (add the username from getpwuid(uid) to the metric set) would be applied to a metric set that already contains the uid and the hostname; thus we have as complete context info as is available. One could argue that cluster name is also needed on some sites that recycle hostnames, but this is always available if needed. 

How one disambiguates things where there are genuine conflicts can only be solved at the analysis scope which includes multiple systems such that there is a conflict. 

Another use case is where the same user has different user names and different uids across different systems and wants to aggregate data across both. A less likely scenario in the national labs, but probably fairly common in a university setting. 

For all three cases of real person userA:
- userA>usernameA=uid1,uid2
- userA>usernameA=uid1,usernameB=uid1
- userA>username=uid1,usernameB=uid2

I claim that the most we can do (and the minimum we should do) at the sampler scope is to provide the locally correct tuple (username,uid,hostname[,optionally cluster name]).

Everything else, if aggregating across disparate identity managements, must be processed at the analysis stage-- there's nothing more possible at the sampler. If sites allow the same username to apply to two different real people, they get what they deserve and will have to define database indices/keys accordingly.
For my environment, users in many if not most cases have no idea what their uid is; they might have two usernames though this would be extremely rare and mostly violates current policies [elevated privilege accounts are not 'the same real user']. They could easily, depending on container uid mappings, have multiple uids to the same username.

One could argue on the same basis that I should also add the option or requirement that if GID is collected then the groupname collection should also be at least attempted.

And for both gid and uid, we need to have a defined string that gets used if the name associated with the gid/uid cannot be currently resolved (yes, this happens from time to time with automated identity management systems). 

Folks writing queries where these aliasing scenarios are possible will need query language such as (name==a || name==b) but i'm pretty sure that's trivial with current tools. It will be up to db admins to understand their site identity management practices and index accordingly. I suspect that for 99+% of the data at my site, we could store only username and never uid and all analysis would proceed smoothly; however, I'd still store uid/username anyway in case the wacko with the closet-cluster comes along with her own passwd file.

The final worrisome use case is that when a user leaves, looking up getpwuid is no longer possible (by design & not in our control). Thus if we don't capture contemporaneously, there's no guarantee that later analyses will be able to construct a correct name:uid mapping.

**Comment by tom95858** [comment]
@baallan, yes, but now we have the converse issue of the name 'tom' matches two uids. I don't see how this solves anything at all. And this advanced syntax where a user has two uids and they want to specify both in the query is already supported.
